<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ sc5mc</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Function reference" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">sc5mc</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.37</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/tanaylab/sc5mc">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-all-functions" class="hasAnchor"><a href="#section-all-functions" class="anchor"></a>All functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="cgdb-class.html">cgdb-class</a></code> </p>
        </td>
        <td><p>cgdb</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cgdb_add_cell_from_df.html">cgdb_add_cell_from_df()</a></code> </p>
        </td>
        <td><p>Add data from data frame to cgdb</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cgdb_add_cell_from_tidy_cpgs.html">cgdb_add_cell_from_tidy_cpgs()</a></code> </p>
        </td>
        <td><p>Add data from tidy_cpgs data frame to cgdb</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cgdb_add_cell_from_tidy_cpgs_dir.html">cgdb_add_cell_from_tidy_cpgs_dir()</a></code> </p>
        </td>
        <td><p>Add data from tidy_cpgs directory to cgdb</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cgdb_add_plate.html">cgdb_add_plate()</a></code> </p>
        </td>
        <td><p>Add sc5mc data of a plate from smat object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cgdb_init.html">cgdb_init()</a></code> </p>
        </td>
        <td><p>Initialize cgdb database</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cgdb_load.html">cgdb_load()</a></code> </p>
        </td>
        <td><p>Load cgdb database from disk</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cgdb_remove_plate.html">cgdb_remove_plate()</a></code> </p>
        </td>
        <td><p>Remove plate from cgdb</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cgdb_save.html">cgdb_save()</a></code> </p>
        </td>
        <td><p>Write cgdb to disk</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cols2ids.html">cols2ids()</a></code> </p>
        </td>
        <td><p>convert column names to ids</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="count_pairs.html">count_pairs()</a></code> </p>
        </td>
        <td><p>Count methylation calls (00,01,10,11) for pairs of cells</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="extract.cgdb.html">extract(<i>&lt;cgdb&gt;</i>)</a></code> </p>
        </td>
        <td><p>Extract data from intervals and cells</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="extract_sparse.html">extract_sparse()</a></code> </p>
        </td>
        <td><p>Extract data from intervals and cells in a sparse format</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="fast_readMM.html">fast_readMM()</a></code> </p>
        </td>
        <td><p>Fast load sparse matrix from MatrixMarket format using data.table::fread</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filter_by_avg.html">filter_by_avg()</a></code> </p>
        </td>
        <td><p>Filter cgdb object by cpgs average methylation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filter_by_cov.html">filter_by_cov()</a></code> </p>
        </td>
        <td><p>Filter cgdb object by cpgs / cells coverage</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filter_intervals.html">filter_intervals()</a></code> </p>
        </td>
        <td><p>Filter CpGs by genomic intervals</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_cgc_trend.html">get_cgc_trend()</a></code> </p>
        </td>
        <td><p>Calculate average methylation per cell given CpG content</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="intervals_cell_cov.html">intervals_cell_cov()</a></code> </p>
        </td>
        <td><p>Calculate number of cells per interval</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="phase_cells.html">phase_cells()</a></code> </p>
        </td>
        <td><p>Phase cells (to G and S) using ratio of coverage in late and early replicating regions</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plpdb-class.html">plpdb-class</a></code> </p>
        </td>
        <td><p>plpdb</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plpdb_add_plate.html">plpdb_add_plate()</a></code> </p>
        </td>
        <td><p>Add plate to plpdb frpm smat reads</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plpdb_init.html">plpdb_init()</a></code> </p>
        </td>
        <td><p>Initialize pileup database</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plpdb_load.html">plpdb_load()</a></code> </p>
        </td>
        <td><p>Load plpdb database from disk</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qc_plot.html">qc_plot()</a></code> </p>
        </td>
        <td><p>Plot QC statistics of cgdb object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sc5mc.html">sc5mc</a></code> </p>
        </td>
        <td><p>sc5mc.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sc5mc.calc_pdiff.html">sc5mc.calc_pdiff()</a></code> </p>
        </td>
        <td><p>calculate pairwise statistics of cells</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sc5mc.calc_pdiff_cor.html">sc5mc.calc_pdiff_cor()</a></code> </p>
        </td>
        <td><p>Calculate pairwise correlation of cells pdiff</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sc5mc.calc_pdiff_rows.html">sc5mc.calc_pdiff_rows()</a></code> </p>
        </td>
        <td><p>calculate pairwise statistics of CpGs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sc5mc.cells_per_umi.html">sc5mc.cells_per_umi()</a></code> </p>
        </td>
        <td><p>Calculates cells per UMI</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sc5mc.cluster_cells.html">sc5mc.cluster_cells()</a></code> </p>
        </td>
        <td><p>Cluster cells</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sc5mc.generate_metacells.html">sc5mc.generate_metacells()</a></code> </p>
        </td>
        <td><p>Genarate metacells from TAD methylation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sc5mc.global_meth_trend.html">sc5mc.global_meth_trend()</a></code> </p>
        </td>
        <td><p>Global methylation trend per cell</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sc5mc.index_plots.html">sc5mc.index_plots()</a></code> </p>
        </td>
        <td><p>plot indexes statistics</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sc5mc.mc2d_force_knn.html">sc5mc.mc2d_force_knn()</a></code> </p>
        </td>
        <td><p>Metacell layout using force directed projection of a low degree mc graph</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sc5mc.plot_cell_marginals.html">sc5mc.plot_cell_marginals()</a></code> </p>
        </td>
        <td><p>plot cell marginals</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sc5mc.plot_cell_pairs_coverage.html">sc5mc.plot_cell_pairs_coverage()</a></code> </p>
        </td>
        <td><p>plot cell pairs coverage</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sc5mc.plot_cells_per_umi.html">sc5mc.plot_cells_per_umi()</a></code> </p>
        </td>
        <td><p>Plots cells per UMI</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sc5mc.plot_cg_pairs_coverage.html">sc5mc.plot_cg_pairs_coverage()</a></code> </p>
        </td>
        <td><p>plot cg pairs coverage</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sc5mc.plot_cor_mat.html">sc5mc.plot_cor_mat()</a></code> </p>
        </td>
        <td><p>plot cell cell correlation matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sc5mc.plot_cpg_marginals.html">sc5mc.plot_cpg_marginals()</a></code> </p>
        </td>
        <td><p>plot cpg marginals</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sc5mc.plot_proj2d.html">sc5mc.plot_proj2d()</a></code> </p>
        </td>
        <td><p>Plot 2d projection of metacells</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sc5mc.qc_plot.html">sc5mc.qc_plot()</a></code> </p>
        </td>
        <td><p>Plot QC statistics of smat object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sc5mcs.calc_pdiff_cor_rows.html">sc5mcs.calc_pdiff_cor_rows()</a></code> </p>
        </td>
        <td><p>calculate pairwise correlation of CpGs pdiff</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="smat.calc_pdiff.html">smat.calc_pdiff()</a></code> </p>
        </td>
        <td><p>calculate pairwise statistics of cells</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="smat.calc_pdiff_cor.html">smat.calc_pdiff_cor()</a></code> </p>
        </td>
        <td><p>calculate pairwise correlation of cells pdiff</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="smat.cell_marginals.html">smat.cell_marginals()</a></code> </p>
        </td>
        <td><p>calculate marginal coverage of cells</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="smat.cell_pairs_marginals.html">smat.cell_pairs_marginals()</a></code> </p>
        </td>
        <td><p>Calculate the joint coverage of all pairs of cells</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="smat.cluster_cells.html">smat.cluster_cells()</a></code> </p>
        </td>
        <td><p>Cluster cells</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="smat.colnames.html">smat.colnames()</a></code> </p>
        </td>
        <td><p>Get colnames of smat object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="smat.cpg_avg_marginals.html">smat.cpg_avg_marginals()</a></code> </p>
        </td>
        <td><p>calculate marginal average methylation of CpGs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="smat.cpg_marginals.html">smat.cpg_marginals()</a></code> </p>
        </td>
        <td><p>calculate marginal coverage of CpGs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="smat.cpg_pairs_marginals.html">smat.cpg_pairs_marginals()</a></code> </p>
        </td>
        <td><p>Calculate the joint coverage of all pairs of CpGs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="smat.filter.html">smat.filter()</a></code> </p>
        </td>
        <td><p>Filter smat object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="smat.filter_by_cov.html">smat.filter_by_cov()</a></code> </p>
        </td>
        <td><p>Filter smat object by cpgs / cells coverage</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="smat.filter_cpgs.html">smat.filter_cpgs()</a></code> </p>
        </td>
        <td><p>Filter smat object by CpG intervals</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="smat.from_bams.html">smat.from_bams()</a></code> </p>
        </td>
        <td><p>Create smat object from mapped reads (bam files)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="smat.from_conf.html">smat.from_conf()</a></code> </p>
        </td>
        <td><p>Load smat object from disk</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="smat.from_df.html">smat.from_df()</a></code> </p>
        </td>
        <td><p>Create smat object from data frame</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="smat.from_tracks.html">smat.from_tracks()</a></code> </p>
        </td>
        <td><p>Create smat object from misha tracks</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="smat.get_tidy_cpgs.html">smat.get_tidy_cpgs()</a></code> </p>
        </td>
        <td><p>Get tidy cpgs of smat object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="smat.info.html">smat.info()</a></code> </p>
        </td>
        <td><p>Print smat object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="smat.join.html">smat.join()</a></code> </p>
        </td>
        <td><p>join two smat objects</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="smat.load.html">smat.load()</a></code> </p>
        </td>
        <td><p>Load smat object from disk</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="smat.multi_join.html">smat.multi_join()</a></code> </p>
        </td>
        <td><p>join multiple smat objects</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="smat.multi_join_from_files.html">smat.multi_join_from_files()</a></code> </p>
        </td>
        <td><p>join multiple smat objects from path</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="smat.plot_cor_mat.html">smat.plot_cor_mat()</a></code> </p>
        </td>
        <td><p>plot cell cell correlation matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="smat.save.html">smat.save()</a></code> </p>
        </td>
        <td><p>Save smat object to disk</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="smat.select.html">smat.select()</a></code> </p>
        </td>
        <td><p>Select cells from smat object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="smat.summarise.html">smat.summarise()</a></code> </p>
        </td>
        <td><p>Summarise groups of CpGs per cell</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="smat.summarise_by_intervals.html">smat.summarise_by_intervals()</a></code> </p>
        </td>
        <td><p>Group CpGs by intervals, and summarise per cell</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="smat.summarise_by_track.html">smat.summarise_by_track()</a></code> </p>
        </td>
        <td><p>Group CpGs by a misha track, and summarise per cell</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="smat.summarise_cpgs.html">smat.summarise_cpgs()</a></code> </p>
        </td>
        <td><p>Summarises CpGs by groups pf cells</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="smat.to_df.html">smat.to_df()</a></code> </p>
        </td>
        <td><p>export smat object to data frame</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="smat.to_marginal_track.html">smat.to_marginal_track()</a></code> </p>
        </td>
        <td><p>Create track from smat marginals</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="summarise.cgdb.html">summarise(<i>&lt;cgdb&gt;</i>)</a></code> </p>
        </td>
        <td><p>Summarise data from intervals and cells</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="summarise_cells.html">summarise_cells()</a></code> </p>
        </td>
        <td><p>Summary of cells per CpG</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="summarise_cpgs.html">summarise_cpgs()</a></code> </p>
        </td>
        <td><p>Summary of CpGs per cell</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="summarise_intervals.html">summarise_intervals()</a></code> </p>
        </td>
        <td><p>Summarise for intervals set</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tidy2smat.html">tidy2smat()</a></code> </p>
        </td>
        <td><p>Create sparse matrix from tidy data frame</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#section-all-functions">All functions</a></li>
    </ul>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Aviezer Lifshitz.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.0.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


